core/src/runtime/hooks.rs

```diff
@@ -0,0 +1,100 @@
+use core::convert::TryInto;
+use std::collections::HashMap;
+
+use k256::ecdsa::{RecoveryId, Signature, VerifyingKey};
+use k256::elliptic_curve::ops::Invert;
+
+use super::Runtime;
+
+type HookId = u32;
+type BoxedHook<'a> = Box<dyn Fn(HookEnv, &[u8]) -> Vec<Vec<u8>> + 'a>;
+
+// Ensure this value is synced with the values in `zkvm/precompiles/src/io.rs`
+/// The file descriptor through which to access `hook_ecrecover`.
+pub const FD_ECRECOVER_HOOK: HookId = 5;
+
+/// A registry of hooks to call, indexed by the file descriptors through which they are accessed.
+pub struct HookRegistry<'a> {
+    /// Table of registered hooks. Prefer using `Runtime::invoke_hook` and
+    /// `HookRegistry::register` over interacting with this field directly.
+    pub table: HashMap<HookId, BoxedHook<'a>>,
+}
+
+impl<'a> HookRegistry<'a> {
+    /// Create a registry with the default hooks.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Create an empty registry.
+    pub fn empty() -> Self {
+        Self {
+            table: Default::default(),
+        }
+    }
+
+    /// Register a hook under a given name.
+    pub fn register(&mut self, name: HookId, hook: BoxedHook<'a>) {
+        self.table.insert(name, hook);
+    }
+}
+
+impl<'a> Default for HookRegistry<'a> {
+    fn default() -> Self {
+        // When `LazyCell` gets stabilized (1.81.0), we can use it to avoid unnecessary allocations.
+        let table = {
+            let entries: Vec<(HookId, BoxedHook)> =
+                vec![(FD_ECRECOVER_HOOK, Box::new(hook_ecrecover))];
+            HashMap::from_iter(entries)
+        };
+
+        Self { table }
+    }
+}
+
+/// Environment that a hook may read from.
+pub struct HookEnv<'a, 'b: 'a> {
+    pub runtime: &'a Runtime<'b>,
+}
+
+pub fn hook_ecrecover(_env: HookEnv, buf: &[u8]) -> Vec<Vec<u8>> {
+    let (sig, msg_hash) = buf.split_at(65);
+    let sig: &[u8; 65] = sig
+        .try_into()
+        .expect("buf should have correct size and offset of bytes for sig");
+    let msg_hash: &[u8; 32] = msg_hash
+        .try_into()
+        .expect("buf should have correct size and offset of bytes for msg_hash");
```

Comment:
nit: this could be a bit cleaner if you just check buf.len() and said something like "expected ecrecover input to have len 97"